src/test/regress/expected/local_table_join.out

```diff
@@ -0,0 +1,1125 @@
+DEBUG:  Wrapping relation "postgres_table" to a subquery: SELECT key, NULL::text AS value, NULL::jsonb AS value_2 FROM local_table_join.postgres_table WHERE true OFFSET 0
+DEBUG:  generating subplan XXX_1 for subquery SELECT key, NULL::text AS value, NULL::jsonb AS value_2 FROM local_table_join.postgres_table WHERE true OFFSET 0
+DEBUG:  Plan XXX query after replacing subqueries and CTEs: SELECT count(*) AS count FROM ((local_table_join.distributed_table JOIN (SELECT intermediate_result.key, intermediate_result.value, intermediate_result.value_2 FROM read_intermediate_result('XXX_1'::text, 'binary'::citus_copy_format) intermediate_result(key integer, value text, value_2 jsonb)) postgres_table USING (key)) JOIN local_table_join.reference_table USING (key))
+ count
+---------------------------------------------------------------------
+     0
+(1 row)
+
+-- partititoned local tables should work as well
+SELECT count(*) FROM distributed_table JOIN local_partitioned_table USING(key);
+DEBUG:  Wrapping relation "local_partitioned_table" to a subquery: SELECT NULL::integer AS key, NULL::text AS value FROM local_table_join.local_partitioned_table WHERE true OFFSET 0
+DEBUG:  generating subplan XXX_1 for subquery SELECT NULL::integer AS key, NULL::text AS value FROM local_table_join.local_partitioned_table WHERE true OFFSET 0
+DEBUG:  Plan XXX query after replacing subqueries and CTEs: SELECT count(*) AS count FROM (local_table_join.distributed_table JOIN (SELECT intermediate_result.key, intermediate_result.value FROM read_intermediate_result('XXX_1'::text, 'binary'::citus_copy_format) intermediate_result(key integer, value text)) local_partitioned_table USING (key))
+ count
+---------------------------------------------------------------------
+     0
+(1 row)
+
+SELECT count(*) FROM reference_table JOIN local_partitioned_table USING(key);
+DEBUG:  Wrapping relation "local_partitioned_table" to a subquery: SELECT NULL::integer AS key, NULL::text AS value FROM local_table_join.local_partitioned_table WHERE true OFFSET 0
+DEBUG:  generating subplan XXX_1 for subquery SELECT NULL::integer AS key, NULL::text AS value FROM local_table_join.local_partitioned_table WHERE true OFFSET 0
+DEBUG:  Plan XXX query after replacing subqueries and CTEs: SELECT count(*) AS count FROM (local_table_join.reference_table JOIN (SELECT intermediate_result.key, intermediate_result.value FROM read_intermediate_result('XXX_1'::text, 'binary'::citus_copy_format) intermediate_result(key integer, value text)) local_partitioned_table USING (key))
+ count
+---------------------------------------------------------------------
+     0
+(1 row)
+
+SELECT count(*) FROM distributed_table JOIN local_partitioned_table USING(key) JOIN reference_table USING (key);
+DEBUG:  Wrapping relation "local_partitioned_table" to a subquery: SELECT NULL::integer AS key, NULL::text AS value FROM local_table_join.local_partitioned_table WHERE true OFFSET 0
```

Comment:
This is definitely wrong, we should possibly have tests that return non-zero results. Here the subquery has all the columns as NULL.

Reply:
The short answer is that we should do the following (or similar):
```C

diff --git a/src/backend/distributed/planner/distributed_planner.c b/src/backend/distributed/planner/distributed_planner.c
index 21de545e2..f4f64dca2 100644
--- a/src/backend/distributed/planner/distributed_planner.c
+++ b/src/backend/distributed/planner/distributed_planner.c
@@ -414,7 +414,7 @@ AdjustPartitioningForDistributedPlanning(List *rangeTableList,
                 * value before and after dropping to the standart_planner.
                 */
                if (rangeTableEntry->rtekind == RTE_RELATION &&
-                       IsCitusTable(rangeTableEntry->relid) &&
+                       //IsCitusTable(rangeTableEntry->relid) &&
                        PartitionedTable(rangeTableEntry->relid))
                {
            
```

The long answer is that for citus tables that are partitioned, we prevent postgres to plan the table as partitioned as it creates relation restrictions per partition. Instead, we treat the partitioned table as a regular tables and let workers handle as partitioned tables. The goal is simplification of the distributed planner.

As we now rely on local tables' restriction information, we should treat local tables similarly, as long as they are part of distributed planning.

Reply:
I actually thought that we could get the restriction with relation ID instead of GetRTEIdentity because for partitioned local tables, `values_list` is already filled hence we skip assigning an identity to partitions of a local table, which means that its identity is a `random` value for us. Then when we get a restriction info we might get the restriction for a partitioned table and there is no restriction hence it has all the columns as `NULL` as there is no required column. 

What is the reason that we don't get the restrictions with `Oid` but with `RTEIdentity`?

Specifically why don't we have such a method:

```c

RelationRestriction *
GetRelationRestrictionForRelation(List* relationRestrictionList, Oid relationId)
{
	List* filteredRelationRestrictionList = NIL;

	RelationRestriction *relationRestriction = NULL;

	foreach_ptr(relationRestriction, relationRestrictionList)
	{
		if (relationRestriction->relationId == relationId) {
			filteredRelationRestrictionList =lappend(filteredRelationRestrictionList, relationRestriction);
		}
	}
	return (RelationRestriction*)linitial(filteredRelationRestrictionList);
}
```

I tried this and it works as expected

Reply:
> What is the reason that we don't get the restrictions with Oid but with RTEIdentity?

OID is not unique, but RTE identity is unique that's  why we cannot rely on Oid. Assume you have the same table twice in the same join clause. Or, more complicated, in a different leaves of a query with subqueries.

We can discuss the distinction in depth on Monday

Reply:
Well then currently there is a bug I think because identity is also not unique for partitioned local tables. For example local_partitioned table and all its partitions might have the same identitiy, I guess we don't want that (because of `values_list` being already non-empty)

Reply:
See the code I shared above. I don't think it was a bug, but it is something that we never did before: Local tables are part of distributed planning. So, it is a new case.

If you let PG, it generates restriction information per partition even if  the query is on the parent. However, by forcing PG to plan a partitioned table as a non-partitioned tables we prevent it. The main thing is that the restrictions on partitions do not provide any help for distributed planning purposes. Now, we should do the same for local partitioned tables as long as we are doing distributed planning. None of the code is robust for partitions to be on the restriction list. So, even if we by-pass this check, we can run into many more edge cases.

Overall, the unique identifier is the `RteIdentity` and that's not something we can easily (or ever?) change. I already had comments around it on the PR, we could never rely on relationId

Reply:
> For example local_partitioned table and all its partitions might have the same identitiy,

Postgres might be copying Rte from parent to partitions, otherwise it seems unlikely that Citus assigns the same values. As noted, we should not have restriction info  partitions, as the code is not considering those.